src/misc/messages.cpp

```diff
@@ -21,227 +21,751 @@
+		while (std::getline(in_file, line)) {
+			++line_number;
+			if (line == ".") {
+				is_text_terminated = true;
+				break;
+			}
+
+			if (is_first_text_line) {
+				is_first_text_line = false;
+			} else {
+				text += "\n";
+			}
+
+			text += line;
+		}
+
+		if (!is_text_terminated) {
+			problem_with_message(name, "message text not terminated");
+			return false;
+		}
+
+		if (text.empty()) {
+			problem_with_message(name, "message text is empty");
+			return false;
+		}
+
+		if (dictionary_translated.contains(name)) {
+			problem_with_message(name, "duplicated message name");
+			return false;
+		}
+
+		dictionary_translated.emplace(name, Message(text));
+
+		auto& translated = dictionary_translated.at(name);
+		if (dictionary_english.contains(name)) {
+			translated.Verify(name, dictionary_english.at(name));
+		}
+	}
+
+	++line_number;
+	if (in_file.bad()) {
+		problem_generic("I/O error");
+		return false;
+	} else if (dictionary_translated.empty()) {
+		problem_generic("file has no content");
+		return false;
+	}
+
+	if (!translation_script) {
+		LOG_WARNING("LOCALE: Translation file did not specify the language script");
+	}
+
+	if (dos.loaded_codepage) {
+		check_code_page();
 	}
-	fclose(mfile);
-	message_file_loaded = true;
+
 	return true;
 }
 
-const char* MSG_Get(const char* requested_name)
+static bool save_messages(const std_fs::path& file_name)
 {
-	const auto it = messages.find(requested_name);
-	if (it != messages.end()) {
-		return it->second.GetRendered();
+	if (file_name.empty()) {
+		return false;
+	}
+
+	constexpr bool RawRequested = true;
+
+	std::ofstream out_file(file_name);
+
+	// Output help line
+	out_file << "// Writing script used in this translation, can be one of:\n";
+	out_file << "// ";
+	bool is_first_script = true;
+	for (const auto& entry : LocaleData::ScriptInfo) {
+		if (is_first_script) {
+			is_first_script = false;
+		} else {
+			out_file << ", ";
+		}
+		out_file << entry.second.script_name;
+	}
+	out_file << "\n";
+
+	// Output script definition
+	std::string script_name = {};
+	if (translation_script) {
+		// Saving translated messages, script is specified
+		assert(LocaleData::ScriptInfo.contains(*translation_script));
+		script_name = LocaleData::ScriptInfo.at(*translation_script).script_name;
+	} else {
+		// Saving English messages, script is always Latin
+		assert(LocaleData::ScriptInfo.contains(Script::Latin));
+		script_name = LocaleData::ScriptInfo.at(Script::Latin).script_name;
+	}
+	if (translation_script || dictionary_translated.empty()) {
+		out_file << KeyScript << script_name;
+	} else {
+		// Script was not specified in the input file
+		out_file << "// " << KeyScript << script_name;
+	}
+	out_file << "\n\n";
+
+	// Save all the messages, in the original order
+	for (const auto& name : message_order) {
+		if (out_file.fail()) {
+			break;
+		}
+
+		std::string text = {};
+		if (dictionary_translated.contains(name)) {
+			text = dictionary_translated.at(name).Get(RawRequested);
```

Comment:
Yeah, so `GetRaw()` would be a lot nicer.

Reply:
OK, changed.